src/App/Fossa/IAT/Fingerprint.hs

```diff
@@ -0,0 +1,39 @@
+module App.Fossa.IAT.Fingerprint (
+  fingerprintContentsRaw,
+) where
+
+import App.Fossa.IAT.Types
+import Control.Carrier.Diagnostics
+import Control.Effect.Lift
+import Crypto.Hash
+import Data.ByteString.Lazy qualified as BL
+import Data.String.Conversion
+import Data.Text
+import Discovery.Walk
+import Effect.ReadFS
+import Path
+
+sha256 :: BL.ByteString -> Digest SHA256
+sha256 = hashlazy
+
+hashFile :: (Has (Lift IO) sig m, Has Diagnostics sig m) => FilePath -> m Text
+hashFile file = do
+  content <- context "Read file" $ sendIO $ BL.readFile file
+  pure . toText . show $ sha256 content
+
+fingerprintRaw :: (Has (Lift IO) sig m, Has Diagnostics sig m) => Path Abs File -> m Fingerprint
+fingerprintRaw file = do
+  fp <- context "Fingerprint file" $ hashFile (toFilePath file)
+  pure (Fingerprint fp)
+
+fingerprintRawMany :: (Has (Lift IO) sig m, Has Diagnostics sig m) => [Path Abs File] -> m [Fingerprint]
+fingerprintRawMany [] = pure []
+fingerprintRawMany (x : xs) = do
+  fp <- fingerprintRaw x
+  fps <- fingerprintRawMany xs
+  pure (fp : fps)
```

Comment:
[nit] This is good. Quick nit to simplify it: you've unknowingly reimplemented `traverse` on lists here.

```hs
traverse :: (Traversable t, Applicative f) => (a -> f b) -> t a -> f (t b)

-- or under the list instance
traverse :: Applicative f => (a -> f b) -> [a] -> f [b]
```

`traverse` is essentially "run an effectful function over everything in the structure, collecting results back into the structure". It's kind of like a specialized for loop in imperative languages, and it turns out that [`for` is another provided function](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Traversable.html#v:for) that's just `traverse` with its arguments reversed. Makes it convenient to do things with a `do` block like:

```hs
example = for someListOfThings $ \oneElement -> do
  foo
  pure bar
```

---

### Intuition-building for `traverse`

1. Note that we get to choose the effect (anything with an `Applicative` instance -- and any `Monad` necessarily has an `Applicative` instance)

The `Maybe` monad/applicative has short-circuiting behavior, so we can use it for, e.g., data validation:
```hs
data Unvalidated = ...
data Validated = ...

validateSingle :: Unvalidated -> Maybe Validated
validateSingle = ...

-- validate everything in the list, returning `Nothing` if any elements are invalid
validateAll :: [Unvalidated] -> Maybe [Validated]
validateAll = traverse validateSingle
```

Similarly the `Either e` monad has short-circuiting behavior, but returns the first error it encounters:

```hs
data ValidationError = ...
data Unvalidated = ...
data Validated = ...

validateSingle :: Unvalidated -> Either ValidationError Validated
validateSingle = ...

-- validate everything in the list, returning `Nothing` if any elements are invalid
validateAll :: [Unvalidated] -> Either ValidationError [Validated]
validateAll = traverse validateSingle
```

and there's another `Applicative` (but not `Monad`) [called `Validation`](https://hackage.haskell.org/package/either-5.0.1.1/docs/Data-Either-Validation.html) that would return _all_ of the validation errors, and not just the first one

2. Traverse is essentially an `fmap` followed by `sequence`

```hs
-- `fmap f` gives us `t (f a)` -- e.g., `fmap putStrLn [1..5] :: [IO ()]`
--
traverse :: (Traversable t, Applicative f) => (a -> f b) -> t a -> f (t b)
traverse f = sequence . fmap f

-- "swap" the position of t and f
sequence :: (Traversable t, Applicative f) => t (f a) -> f (t a)
sequence = ...
```

3. It's common to not care about the results of the effectful action, like in a typical imperative language for loop

For this reason, there are `traverse_` and `for_` variants that "ignore the results":
```hs
traverse_ :: (Foldable t, Applicative f) => (a -> f b) -> t a -> f ()
for_ :: (Foldable t, Applicative f) => t a -> (a -> f b) -> f ()

-- e.g.,

```hs
main :: IO ()
main = traverse_ print [1..100]
```

Reply:
This was super helpful- and also gives me the building block for the future. I was trying to use just `fmap` and was running into the issue where the monad was inside the result (`[IO ()]`) and couldn't figure out how to get it out, so fell back to recursion- now I know `sequence` is what gets it out, and to do both it's just `traverse`. Thanks!